File: java/leetcode_grind/Day453.java
package leetcode_grind;

public class Day453 {
    // https://leetcode.com/problems/paint-fence/description
    static class Solution1 {
        public int numWays(int n, int k) {
            int[] dp = new int[n + 1];
            dp[0] = k;
            dp[1] = k * k;

            for (int i = 2; i < n; i++) {
                dp[i] = (k - 1) * (dp[i - 1] + dp[i - 2]);
            }

            return dp[n - 1];
        }
    }
}

File: src/leetcode_grind/day_453.rs
// https://leetcode.com/problems/perfect-squares/description
pub fn num_squares(n: i32) -> i32 {
    use std::collections::HashSet;

    let mut squares = vec![];
    let mut i = 1;

    while i * i <= n {
        squares.push(i * i);
        i += 1;
    }

    let mut queue = HashSet::new();
    queue.insert(n);

    let mut lvl = 1;
    while queue.len() > 0 {
        let mut nxt = HashSet::new();
        for v in queue {
            for &s in &squares {
                if v < s {
                    break;
                } else if v == s {
                    return lvl;
                } else {
                    nxt.insert(v - s);
                }
            }
        }
        queue = nxt;
        lvl += 1;
    }

    lvl
}
